piptools/resolver.py

```diff
@@ -448,3 +478,164 @@ def reverse_dependencies(
             ireq for ireq in ireqs if not (ireq.editable or is_url_requirement(ireq))
         ]
         return self.dependency_cache.reverse_dependencies(non_editable)
+
+
+class Resolver(BaseResolver):
+    """
+    Uses pip's 2020 resolver.
+    """
+
+    def __init__(
+        self,
+        constraints: Iterable[InstallRequirement],
+        existing_constraints: typing.Dict[str, InstallRequirement],
+        repository: BaseRepository,
+        allow_unsafe: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        self.constraints = list(constraints)
+        self.repository = repository
+        self.allow_unsafe = allow_unsafe
+
+        options = self.options = self.repository.options
+        self.session = self.repository.session
+        self.finder = self.repository.finder
+        self.command = self.repository.command
+        self.unsafe_constraints: Set[InstallRequirement] = set()
+
+        self.existing_constraints = existing_constraints
+        self._constraints_map = {key_from_ireq(ireq): ireq for ireq in constraints}
+
+        if "2020-resolver" not in options.features_enabled:
+            raise PipToolsError("2020 resolver feature must be enabled.")
+
+        # Make sure there is no enabled legacy resolver
+        options.deprecated_features_enabled = remove_value(
+            options.deprecated_features_enabled, "legacy-resolver"
+        )
+
+    def resolve(self, max_rounds: int = 10) -> Set[InstallRequirement]:
+        with get_requirement_tracker() as req_tracker, global_tempdir_manager(), indent_log(), update_env_context_manager(  # noqa: E501
+            PIP_EXISTS_ACTION="i"
+        ):
+            # Mark direct/primary/user_supplied packages
+            for ireq in self.constraints:
+                ireq.user_supplied = True
+
+            # Pass compiled requirements from `requirements.txt` as constraints to resolver
+            existing_constraints = list(self.existing_constraints.values())
+            for ireq in existing_constraints:
+                ireq.extras = set()  # pip does not support extras in constraints
+                ireq.constraint = True
+                ireq.user_supplied = False
+                ireq._is_existing_pin = True
```

Comment:
How about just doing
```suggestion
                ireq._is_existing_pin = True
                self.constraints.append(ireq)
```
and you wouldn't need to maintain an extra variable

Reply:
Addressed in 7035524